src/shogun/mathematics/linalg/internal/implementation/ElementwiseUnaryOperation.h

```diff
@@ -0,0 +1,198 @@
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are those
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ */
+
+#ifndef ELEMENTWISE_OPERATION_H_
+#define ELEMENTWISE_OPERATION_H_
+
+#include <shogun/lib/config.h>
+#include <shogun/lib/SGMatrix.h>
+
+#ifdef HAVE_VIENNACL
+#include <shogun/lib/GPUMatrix.h>
+#include <shogun/mathematics/linalg/internal/opencl_config.h>
+#include <shogun/mathematics/linalg/internal/opencl_util.h>
+#include <shogun/mathematics/linalg/internal/implementation/operations/opencl_operation.h>
+#endif // HAVE_VIENNACL
+
+#include <algorithm>
+#include <type_traits>
+
+namespace shogun
+{
+
+namespace linalg
+{
+
+namespace implementation
+{
+
+/**
+ * @brief Template struct elementwise_unary_operation. This struct is specialized for
+ * computing element-wise operations for both matrices and vectors of CPU
+ * (SGMatrix/SGVector) or GPU (CGPUMatrix/CGPUVector).
+ */
+template <enum Backend, class Operand, class ReturnType, class UnaryOp>
+struct elementwise_unary_operation
+{
+};
+
+/**
+ * @brief Specialization for elementwise_unary_operation with NATIVE backend.
+ * The operand types MUST be of CPU types (SGMatrix/SGVector).
+ */
+template <class Operand, class ReturnType, class UnaryOp>
+struct elementwise_unary_operation<Backend::NATIVE, Operand, ReturnType, UnaryOp>
+{
+	/** The scalar type of the operand */
+	using T = typename Operand::Scalar;
+
+	/** The scalar type of the result */
+	using ST = typename ReturnType::Scalar;
+
+#ifdef HAVE_VIENNACL
+	/** Ensure that this struct is not being instantiated with any GPU operand types */
+	static_assert(std::is_same<SGMatrix<T>, Operand>::value
+			|| std::is_same<SGVector<T>, Operand>::value,
+			"NATIVE backend not allowed for GPU operands! Use SGMatrix/SGVector "
```

Comment:
We have a dedicated ViennaCL build in buildbot. But that just sets the only the global backends to ViennaCL. I think we should get rid of global ViennaCL backend setups, because when that's done, one would take a massive performance hit due to all the data transfer in all the operations. Backends should always be specified explicitly. This static assert is an attempt to prevent from doing the such unintentional things.